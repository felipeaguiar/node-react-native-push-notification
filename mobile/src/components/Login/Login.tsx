import React from 'react';
import { StyleSheet, View, Button, TextInput, Text } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import AsyncStorage from '@react-native-community/async-storage';
import { useDispatch, useSelector } from 'react-redux';
import { Formik } from 'formik';
import * as yup from 'yup';
import api, { isApiError } from '../../api';
import { jwtUtil } from '../../util/jwt';
import { loadingAction, savaTokenAction, failAction } from '../../store/ducks/auth/actions';
import { hasLoginError, getLoginErrorMessage, isLoading } from '../../store/ducks/auth/selectors';

export default function Login() {
  const dispatch = useDispatch();
  const navigation = useNavigation();

  const loading = useSelector(isLoading);
  const hasError = useSelector(hasLoginError);
  const errorMessage = useSelector(getLoginErrorMessage);

  const initialValues = {
    email: '',
    senha: ''
  };

  const validationSchema = yup.object().shape({
    email: yup.string()
      .email('O valor deve ser um email válido')
      .required('E-mail é um campo obrigatório.'),
    senha: yup.string()
      .min(6, 'A senha deve ter pelo menos 6 caracteres')
      .required('Senha é um campo obrigatório.')
  });

  async function setPlayerId() {
    const player = await AsyncStorage.getItem('player') as string;
    await api.put(`usuario/player-id/${player}`);
  }

  async function onSubmit(auth: typeof initialValues) {
    dispatch(loadingAction());

    try {
      const response = await api.post('auth/login', auth);

      const token = response.data.token;
      const jwt = jwtUtil.decode(token);

      if (!jwt) {
        throw new Error();
      }

      await AsyncStorage.setItem('token', token);
      dispatch(savaTokenAction(token, jwt));

      await setPlayerId();
    } catch (error) {
      let message = 'Erro ao efetuar login';

      if (isApiError(error)) {
        message = error.response.data.error.message;
      }

      await AsyncStorage.removeItem('token');
      dispatch(failAction(message));
    }
  }

  function onRegisterClick() {
    navigation.navigate('Register');
  }

  return (
    <Formik initialValues={initialValues} onSubmit={onSubmit} validationSchema={validationSchema}>
        {({ handleChange, handleBlur, handleSubmit, values, errors, touched }) => (
        <View style={styles.container}>

          { hasError && <Text>{errorMessage}</Text> }

          <TextInput
            placeholder="E-mail"
            autoCapitalize="none"
            onChangeText={handleChange('email')}
            onBlur={handleBlur('email')}
            value={values.email}
          />

          { errors.email && touched.email && <Text>{errors.email}</Text> }

          <TextInput
            placeholder="Senha"
            secureTextEntry
            onChangeText={handleChange('senha')}
            onBlur={handleBlur('senha')}
            value={values.senha}
          />

          { errors.senha && touched.senha && <Text>{errors.senha}</Text> }

          <Button title="Login" disabled={loading} onPress={handleSubmit} />

          <Button title="Criar Conta" disabled={loading} onPress={onRegisterClick} />
        </View>
      )}
    </Formik>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center'
  }
});
