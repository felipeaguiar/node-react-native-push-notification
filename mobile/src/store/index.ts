import { createStore, combineReducers} from 'redux';
import { authReducer } from './ducks/auth/reducer';

export const rootReducer = combineReducers({
  auth: authReducer
});

const store = createStore(rootReducer);

export type RootState = ReturnType<typeof rootReducer>
export type Store = typeof store;

export default store;
