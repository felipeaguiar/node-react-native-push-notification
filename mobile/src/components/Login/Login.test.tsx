import '@testing-library/jest-native/extend-expect';
import React from 'react';
import { fireEvent, waitFor } from 'react-native-testing-library';
import AsyncStorage from '@react-native-community/async-storage';
import Login from './Login';
import { loginMock } from '../../test/http-mock';
import store from '../../store';
import { renderWithRedux } from '../../test/reduxMock';

it('login snapshot', () => {
  expect.hasAssertions();

  const { toJSON } = renderWithRedux(<Login />);
  expect(toJSON()).toMatchSnapshot();
});

it('login teste de integração com sucesso', async () => {
  expect.hasAssertions();

  loginMock('felipe@mail.com', '123456');

  const { toJSON, getByPlaceholder, getByText, queryByText } = renderWithRedux(<Login />, { store } );

  const email = getByPlaceholder('E-mail');
  const senha = getByPlaceholder('Senha');
  const login = getByText('Login');

  fireEvent.changeText(email, 'felipe@mail.com');
  fireEvent.changeText(senha, '123456');

  fireEvent.press(login);

  await waitFor(() => expect(login).toBeEnabled());

  const token = await AsyncStorage.getItem('token');
  const alert = queryByText('Usuário ou senha inválidos');

  expect(toJSON()).toMatchSnapshot();
  expect(email).toHaveProp('value', 'felipe@mail.com');
  expect(senha).toHaveProp('value', '');
  expect(alert).toBeFalsy();
  expect(token).not.toBeNull();
  expect(token).toBe(store.getState().auth.token);
});

it('login teste de integração com falha', async () => {
  expect.hasAssertions();

  loginMock('felipe@mail.com', '654321');

  const { toJSON, getByPlaceholder, getByText, queryByText } = renderWithRedux(<Login />, { store } );

  const email = getByPlaceholder('E-mail');
  const senha = getByPlaceholder('Senha');
  const login = getByText('Login');

  fireEvent.changeText(email, 'felipe@mail.com');
  fireEvent.changeText(senha, '654321');

  fireEvent.press(login);

  await waitFor(() => expect(login).toBeEnabled());

  const token = await AsyncStorage.getItem('token');
  const alert = queryByText('Usuário ou senha inválidos');

  expect(toJSON()).toMatchSnapshot();
  expect(email).toHaveProp('value', 'felipe@mail.com');
  expect(senha).toHaveProp('value', '');
  expect(alert).toBeTruthy();
  expect(token).toBeNull();
  expect(token).toBe(store.getState().auth.token);
});
