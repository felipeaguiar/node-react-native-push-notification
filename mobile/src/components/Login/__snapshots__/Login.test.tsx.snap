// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="E-mail"
    rejectResponderTermination={true}
    underlineColorAndroid="transparent"
    value=""
  />
  <TextInput
    allowFontScaling={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Senha"
    rejectResponderTermination={true}
    secureTextEntry={true}
    underlineColorAndroid="transparent"
    value=""
  />
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
          Object {},
        ]
      }
    >
      <Text
        disabled={true}
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
            Object {
              "color": "#cdcdcd",
            },
          ]
        }
      >
        Login
      </Text>
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
          Object {},
        ]
      }
    >
      <Text
        disabled={true}
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
            Object {
              "color": "#cdcdcd",
            },
          ]
        }
      >
        Criar Conta
      </Text>
    </View>
  </View>
</View>
`;

exports[`login teste de integração com falha 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text>
    Usuário ou senha inválidos
  </Text>
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="E-mail"
    rejectResponderTermination={true}
    underlineColorAndroid="transparent"
    value="felipe@mail.com"
  />
  <TextInput
    allowFontScaling={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Senha"
    rejectResponderTermination={true}
    secureTextEntry={true}
    underlineColorAndroid="transparent"
    value="654321"
  />
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        disabled={false}
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Login
      </Text>
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        disabled={false}
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Criar Conta
      </Text>
    </View>
  </View>
</View>
`;

exports[`login teste de integração com sucesso 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="E-mail"
    rejectResponderTermination={true}
    underlineColorAndroid="transparent"
    value="felipe@mail.com"
  />
  <TextInput
    allowFontScaling={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Senha"
    rejectResponderTermination={true}
    secureTextEntry={true}
    underlineColorAndroid="transparent"
    value="123456"
  />
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        disabled={false}
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Login
      </Text>
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        disabled={false}
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Criar Conta
      </Text>
    </View>
  </View>
</View>
`;
